source/common/http/http2/codec_impl.cc

```diff
@@ -90,7 +90,19 @@ void ConnectionImpl::StreamImpl::encode100ContinueHeaders(const HeaderMap& heade
 
 void ConnectionImpl::StreamImpl::encodeHeaders(const HeaderMap& headers, bool end_stream) {
   std::vector<nghttp2_nv> final_headers;
-  buildHeaders(final_headers, headers);
+
+  if (Http::Utility::isUpgrade(headers)) {
+    HeaderMapImpl modified_headers(headers);
+    upgrade_type_ = headers.Upgrade()->value().c_str();
+    if (headers.Status()) {
```

Comment:
nit: it feels a bit odd that we are handling both request/response at the base stream layer and I wonder if we should have request/response stream overrides to make this code more clear? I don't feel very strongly about this but throwing it out there. Same comment in the decode path.

Reply:
Moving the header munging out makes for a lot of duplicate code which I'd prefer to avoid.  I could either move the complexity (status check) to the utility or have a Server/Client wrapper for transforms.  I'll start with the latter and see what folks think of it.